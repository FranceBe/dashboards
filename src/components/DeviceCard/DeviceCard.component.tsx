import ConfirmationNumberIcon from '@material-ui/icons/ConfirmationNumber'
import ContactlessIcon from '@material-ui/icons/Contactless'
import FlashOnIcon from '@material-ui/icons/FlashOn'
import LinkIcon from '@material-ui/icons/Link'
import VisibilityIcon from '@material-ui/icons/Visibility'
import ConnectionType from 'components/ConnectionType'
import {
  CardContainer,
  Hr,
  InfoContainer,
  Label,
  LineContainer,
  Value,
} from 'components/DeviceCard/deviceCard.style'
import Status from 'components/Status'
import moment from 'moment'
import React from 'react'
import { DashboardDeviceable } from 'types/dashboardDevice'

export const DeviceCard: React.FC<{ device: DashboardDeviceable }> = ({ device }) => {
  return (
    <CardContainer data-testid={'device-card'}>
      <LineContainer>
        <InfoContainer>
          <Label>Serial Number :</Label>
          <Value>{device.serial_number}</Value>
        </InfoContainer>
        <InfoContainer>
          <ConnectionType connection_type={device.connection_type} />
          <Label>Connection Type :</Label>
          <Value>{device.connection_type}</Value>
        </InfoContainer>
        <Status status={device.status} />
      </LineContainer>
      <Hr />
      <LineContainer>
        <InfoContainer>
          <FlashOnIcon className={'small-icon'} />
          <Label>Voltage :</Label>
          <Value>{device.voltage}</Value>
        </InfoContainer>
        <InfoContainer>
          <VisibilityIcon className={'small-icon'} />
          <Label>Last seen at :</Label>
          <Value>{moment(device.last_seen_at).format('MMMM Do YYYY, h:mm:ss a')}</Value>
        </InfoContainer>
      </LineContainer>
      <LineContainer>
        <InfoContainer>
          <ContactlessIcon className={'small-icon'} />
          <Label>Mac wifi :</Label>
          <Value>{device.mac_wifi}</Value>
        </InfoContainer>
        <InfoContainer>
          <ConfirmationNumberIcon className={'small-icon'} />
          <Label>Sim id :</Label>
          <Value>{device.sim_id}</Value>
        </InfoContainer>
      </LineContainer>
      <LineContainer>
        <InfoContainer>
          <LinkIcon className={'small-icon'} />
          <Label>Url :</Label>
          <Value>
            <a href={device.url}>{device.url}</a>
          </Value>
        </InfoContainer>
      </LineContainer>
    </CardContainer>
  )
}

export default DeviceCard
