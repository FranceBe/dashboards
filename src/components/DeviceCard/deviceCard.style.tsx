import styled from 'styled-components'
import { palette, spaces } from 'styles/variables'

export const CardContainer = styled.div`
  background-color: ${palette.white};
  display: flex;
  flex-direction: column;
  padding: ${spaces.regular};
  border-radius: ${spaces.small};
  .small-icon {
    margin-right: ${spaces.small};
  }
`
export const Hr = styled.div`
  height: 1px;
  width: 100%;
  border-bottom: 1px solid ${palette.blue_primary};
`
export const LineContainer = styled.div`
  width: 100%;
  display: flex;
  justify-content: space-between;
  align-items: center;
  margin-bottom: ${spaces.medium};
  margin-top: ${spaces.medium};
`

export const ColumnContainer = styled.div`
  display: flex;
  flex-direction: column;
`
export const InfoContainer = styled.div`
  display: flex;
  align-items: center;
`

export const Label = styled.div`
  color: ${palette.blue_primary};
`

export const Value = styled.div`
  margin-left: ${spaces.small};
`
