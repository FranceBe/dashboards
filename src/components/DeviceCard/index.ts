import DeviceCardComponent, { DeviceCard } from 'components/DeviceCard/DeviceCard.component'

export { DeviceCard }

export default DeviceCardComponent
