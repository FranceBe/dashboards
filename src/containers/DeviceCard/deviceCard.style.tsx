import styled from 'styled-components'
import { elements, fontSizes, palette, spaces } from 'styles/variables'

export const CardContainer = styled.div`
  background-color: ${palette.white};
  display: flex;
  flex-direction: column;
  padding: ${spaces.regular} ${spaces.x_large};
  box-shadow: ${elements.box_shadow};
  border: 1px solid ${palette.grey_primary};
  border-radius: ${spaces.small};
  font-size: ${fontSizes.medium};
  .small-icon {
    margin-right: ${spaces.small};
    color: ${palette.blue_secondary};
  }
  .status-container {
    width: 100px;
    height: 100px;
    margin: 0;
    svg {
      width: 100%;
      height: 100%;
    }
  }
`
export const Hr = styled.div`
  height: 1px;
  width: 100%;
  border-bottom: 1px solid ${palette.blue_secondary};
`
export const LineContainer = styled.div`
  width: 100%;
  display: flex;
  flex-wrap: wrap;
  justify-content: space-between;
  align-items: center;
  margin-bottom: ${spaces.medium};
  margin-top: ${spaces.medium};
`

export const ColumnContainer = styled.div`
  display: flex;
  flex-direction: column;
  flex-wrap: wrap;
  justify-content: center;
  .info-container {
    margin-bottom: ${spaces.medium};
  }
  &.status-column {
    justify-content: center;
    align-items: center;
  }
`
export const InfoContainer = styled.div`
  display: flex;
  align-items: center;
`

export const Label = styled.div`
  color: ${palette.blue_secondary};
  text-decoration: underline;
`

export const Value = styled.div`
  margin-left: ${spaces.small};
`
