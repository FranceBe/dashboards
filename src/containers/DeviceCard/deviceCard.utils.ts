// To display Labels for each category this function convert
// a snake_case string into a more lisible string starting with a capital letter and with spaces instead of _
export const convertStringForTitle = (string: string): string =>
  string.charAt(0).toUpperCase() + string.slice(1).replace(/_/g, ' ')
