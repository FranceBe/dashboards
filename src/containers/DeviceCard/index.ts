import DeviceCardComponent, { DeviceCard } from 'containers/DeviceCard/DeviceCard.container'

export { DeviceCard }

export default DeviceCardComponent
