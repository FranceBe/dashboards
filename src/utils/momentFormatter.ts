import moment from 'moment'

export const momentFormatter = (date: string): string =>
  moment(date).format('MMMM Do YYYY, h:mm:ss a')
