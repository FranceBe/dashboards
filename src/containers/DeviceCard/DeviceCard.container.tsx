import ConfirmationNumberIcon from '@material-ui/icons/ConfirmationNumber'
import ContactlessIcon from '@material-ui/icons/Contactless'
import FlashOnIcon from '@material-ui/icons/FlashOn'
import LinkIcon from '@material-ui/icons/Link'
import SettingsEthernetIcon from '@material-ui/icons/SettingsEthernet'
import SignalCellularAltIcon from '@material-ui/icons/SignalCellularAlt'
import VisibilityIcon from '@material-ui/icons/Visibility'
import WifiIcon from '@material-ui/icons/Wifi'
import Status from 'components/Status'
import {
  CardContainer,
  ColumnContainer,
  Hr,
  InfoContainer,
  Label,
  LineContainer,
  Value,
} from 'containers/DeviceCard/deviceCard.style'
import { convertStringForTitle } from 'containers/DeviceCard/deviceCard.utils'
import React, { ReactElement } from 'react'
import { DashboardDeviceable } from 'types/dashboardDevice'
import { momentFormatter } from 'utils/momentFormatter'

export const DeviceCard: React.FC<{ device: DashboardDeviceable }> = ({ device }) => {
  const iconByType: Record<'ethernet' | 'wifi' | 'cellular', ReactElement> = {
    cellular: <SignalCellularAltIcon data-testid={'icon-cellular'} className={'small-icon'} />,
    ethernet: <SettingsEthernetIcon data-testid={'icon-ethernet'} className={'small-icon'} />,
    wifi: <WifiIcon data-testid={'icon-wifi'} className={'small-icon'} />,
  }
  return (
    <CardContainer data-testid={'device-card'}>
      <LineContainer>
        <InfoContainer>
          <Label>{convertStringForTitle('serial_number')}</Label>
          {' : '}
          <Value>{device.serial_number}</Value>
        </InfoContainer>
        <ColumnContainer className={'status-column'}>
          <Status status={device.status} />
          {convertStringForTitle(device.status)}
        </ColumnContainer>
      </LineContainer>
      <Hr />
      <LineContainer>
        <ColumnContainer>
          <InfoContainer className={'info-container'}>
            <VisibilityIcon className={'small-icon'} />
            <Label>{convertStringForTitle('last_seen_at')}</Label>
            {' : '}
            <Value>{momentFormatter(device.last_seen_at)}</Value>
          </InfoContainer>
          <InfoContainer className={'info-container'}>
            {iconByType[device.connection_type]}
            <Label>{convertStringForTitle('connection_type')}</Label>
            {' : '}
            <Value>{device.connection_type}</Value>
          </InfoContainer>
          <InfoContainer className={'info-container'}>
            <FlashOnIcon className={'small-icon'} />
            <Label>{convertStringForTitle('voltage')}</Label>
            {' : '}
            <Value>{device.voltage}</Value>
          </InfoContainer>
        </ColumnContainer>
        <ColumnContainer>
          <InfoContainer className={'info-container'}>
            <ContactlessIcon className={'small-icon'} />
            <Label>{convertStringForTitle('mac_wifi')}</Label>
            {' : '}
            <Value>{device.mac_wifi}</Value>
          </InfoContainer>
          <InfoContainer className={'info-container'}>
            <ConfirmationNumberIcon className={'small-icon'} />
            <Label>{convertStringForTitle('sim_id')}</Label>
            {' : '}
            <Value>{device.sim_id}</Value>
          </InfoContainer>
          <InfoContainer className={'info-container'}>
            <LinkIcon className={'small-icon'} />
            <Label>{convertStringForTitle('url')}</Label>
            {' : '}
            <Value>
              <a href={device.url}>{device.url}</a>
            </Value>
          </InfoContainer>
        </ColumnContainer>
      </LineContainer>
    </CardContainer>
  )
}

export default DeviceCard
