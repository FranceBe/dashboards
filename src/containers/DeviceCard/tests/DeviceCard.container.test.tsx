import '@testing-library/jest-dom/extend-expect'
import 'jest-styled-components'

import { render, screen } from '@testing-library/react'
import DeviceCard from 'containers/DeviceCard/index'
import React from 'react'
import { DashboardDeviceable } from 'types/dashboardDevice'
import { momentFormatter } from 'utils/momentFormatter'

describe('DeviceCard', () => {
  const device: DashboardDeviceable = {
    connection_type: 'cellular',
    last_seen_at: '2021-09-27T14:59:09.021426',
    mac_wifi: '44:20:82:3c:fd:e6',
    serial_number: 'device_1',
    sim_id: 73383683,
    status: 'disconnected',
    url: 'http://backend:8010/device/1',
    voltage: 8.167522699950393,
  }
  it('should match snapshot', () => {
    const { container } = render(<DeviceCard device={device} />)
    expect(container.firstChild).toMatchSnapshot()
  })
  it('should display data info of the device', () => {
    render(<DeviceCard device={device} />)
    expect(screen.getByText('Serial number')).toBeInTheDocument()
    expect(screen.getByText('device_1')).toBeInTheDocument()
    expect(screen.getByText('Connection type')).toBeInTheDocument()
    expect(screen.getByText('cellular')).toBeInTheDocument()
    expect(screen.getByText('Voltage')).toBeInTheDocument()
    expect(screen.getByText('8.167522699950393')).toBeInTheDocument()
    expect(screen.getByText('Last seen at')).toBeInTheDocument()
    expect(screen.getByText(momentFormatter('2021-09-27T14:59:09.021426'))).toBeInTheDocument()
    expect(screen.getByText('Mac wifi')).toBeInTheDocument()
    expect(screen.getByText('44:20:82:3c:fd:e6')).toBeInTheDocument()
    expect(screen.getByText('Sim id')).toBeInTheDocument()
    expect(screen.getByText('73383683')).toBeInTheDocument()
    expect(screen.getByText('Url')).toBeInTheDocument()
    expect(screen.getByText('http://backend:8010/device/1')).toBeInTheDocument()
  })
})
