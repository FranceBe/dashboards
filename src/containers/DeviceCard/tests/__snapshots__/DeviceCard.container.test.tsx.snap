// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeviceCard should match snapshot 1`] = `
.c6 {
  background-color: #eb4034;
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
  border-radius: 4px;
  padding: 8px;
  font-size: 0.75rem;
  box-shadow: 0 5px 25px 0 rgba(0,0,45,0.1),0 2px 6px 0 rgba(0,0,0,0.04), 0 16px 24px 0 rgba(0,0,46,0.06);
}

.c0 {
  background-color: #ffffff;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  padding: 8px 32px;
  box-shadow: 0 5px 25px 0 rgba(0,0,45,0.1),0 2px 6px 0 rgba(0,0,0,0.04), 0 16px 24px 0 rgba(0,0,46,0.06);
  border: 1px solid #343536;
  border-radius: 4px;
  font-size: 1.25rem;
}

.c0 .small-icon {
  margin-right: 4px;
  color: #0061A9;
}

.c0 .status-container {
  width: 100px;
  height: 100px;
  margin: 0;
}

.c0 .status-container svg {
  width: 100%;
  height: 100%;
}

.c7 {
  height: 1px;
  width: 100%;
  border-bottom: 1px solid #0061A9;
}

.c1 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin-bottom: 16px;
  margin-top: 16px;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c5 .info-container {
  margin-bottom: 16px;
}

.c5.status-column {
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c3 {
  color: #0061A9;
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c4 {
  margin-left: 4px;
}

<div
  class="c0"
  data-testid="device-card"
>
  <div
    class="c1"
  >
    <div
      class="c2"
    >
      <div
        class="c3"
      >
        Serial number
      </div>
       : 
      <div
        class="c4"
      >
        device_1
      </div>
    </div>
    <div
      class="c5 status-column"
    >
      <div
        class="c6 status-container"
        data-testid="status-container"
        status="disconnected"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root"
          data-testid="signal-off"
          focusable="false"
          title="disconnected"
          viewBox="0 0 24 24"
        >
          <path
            d="M21 1l-8.59 8.59L21 18.18V1zM4.77 4.5L3.5 5.77l6.36 6.36L1 21h17.73l2 2L22 21.73 4.77 4.5z"
          />
        </svg>
      </div>
      Disconnected
    </div>
  </div>
  <div
    class="c7"
  />
  <div
    class="c1"
  >
    <div
      class="c5"
    >
      <div
        class="c2 info-container"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root small-icon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 4.5C7 4.5 2.73 7.61 1 12c1.73 4.39 6 7.5 11 7.5s9.27-3.11 11-7.5c-1.73-4.39-6-7.5-11-7.5zM12 17c-2.76 0-5-2.24-5-5s2.24-5 5-5 5 2.24 5 5-2.24 5-5 5zm0-8c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3z"
          />
        </svg>
        <div
          class="c3"
        >
          Last seen at
        </div>
         : 
        <div
          class="c4"
        >
          September 27th 2021, 2:59:09 pm
        </div>
      </div>
      <div
        class="c2 info-container"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root small-icon"
          data-testid="icon-cellular"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M17 4h3v16h-3zM5 14h3v6H5zm6-5h3v11h-3z"
          />
        </svg>
        <div
          class="c3"
        >
          Connection type
        </div>
         : 
        <div
          class="c4"
        >
          cellular
        </div>
      </div>
      <div
        class="c2 info-container"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root small-icon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 2v11h3v9l7-12h-4l4-8z"
          />
        </svg>
        <div
          class="c3"
        >
          Voltage
        </div>
         : 
        <div
          class="c4"
        >
          8.167522699950393
        </div>
      </div>
    </div>
    <div
      class="c5"
    >
      <div
        class="c2 info-container"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root small-icon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zM8.46 14.45l-1.36-.62c.28-.61.41-1.24.4-1.86-.01-.63-.14-1.24-.4-1.8l1.36-.63c.35.75.53 1.56.54 2.4.01.86-.17 1.7-.54 2.51zm3.07 1.56l-1.3-.74c.52-.92.78-1.98.78-3.15 0-1.19-.27-2.33-.8-3.4l1.34-.67c.64 1.28.96 2.65.96 4.07 0 1.43-.33 2.74-.98 3.89zm3.14 1.32l-1.35-.66c.78-1.6 1.18-3.18 1.18-4.69 0-1.51-.4-3.07-1.18-4.64l1.34-.67c.9 1.78 1.34 3.56 1.34 5.31 0 1.74-.44 3.54-1.33 5.35z"
          />
        </svg>
        <div
          class="c3"
        >
          Mac wifi
        </div>
         : 
        <div
          class="c4"
        >
          44:20:82:3c:fd:e6
        </div>
      </div>
      <div
        class="c2 info-container"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root small-icon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M22 10V6c0-1.11-.9-2-2-2H4c-1.1 0-1.99.89-1.99 2v4c1.1 0 1.99.9 1.99 2s-.89 2-2 2v4c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2v-4c-1.1 0-2-.9-2-2s.9-2 2-2zm-9 7.5h-2v-2h2v2zm0-4.5h-2v-2h2v2zm0-4.5h-2v-2h2v2z"
          />
        </svg>
        <div
          class="c3"
        >
          Sim id
        </div>
         : 
        <div
          class="c4"
        >
          73383683
        </div>
      </div>
      <div
        class="c2 info-container"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root small-icon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M3.9 12c0-1.71 1.39-3.1 3.1-3.1h4V7H7c-2.76 0-5 2.24-5 5s2.24 5 5 5h4v-1.9H7c-1.71 0-3.1-1.39-3.1-3.1zM8 13h8v-2H8v2zm9-6h-4v1.9h4c1.71 0 3.1 1.39 3.1 3.1s-1.39 3.1-3.1 3.1h-4V17h4c2.76 0 5-2.24 5-5s-2.24-5-5-5z"
          />
        </svg>
        <div
          class="c3"
        >
          Url
        </div>
         : 
        <div
          class="c4"
        >
          <a
            href="http://localhost:8010/device/1"
          >
            http://localhost:8010/device/1
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;
